docs(gpus): AS-561 Document gpu workloads (#328)
voxel51/fiftyone-teams-app-deploy: docker/docs/configuring-gpu-workloads.md

<!-- tocstop -->

## Overview

Many machine learning applications utilize
GPU hardware for intensive computations.
The FiftyOne Enterprise docker compose files allow users to schedule containers on
GPU-enabled nodes using a service's
[`deploy.resource.reservation.devices`][compose-deploy-resources].

The below will show an example deploying a GPU-enabled container via docker
compose by following their
[documentation][compose-gpu-how-to]
in a FiftyOne Enterprise context.

## Utilizing GPUs For Delegated Operations

### Prerequisites

This example assumes you have a docker compose node with GPU devices available
and you have followed Docker's
[GPU prerequisites][compose-gpu-resources].

### Deploying GPU-enabled Delegated Operator Containers

We will configure the
[delegated operators](./configuring-delegated-operators.md)
with GPUs.

In your `compose.delegated-operators.yaml`, under `.services.teams-do`,
add the `deploy.resource.reservation.devices` configuration.

The below will deploy a CPU-based delegated operator (`teams-do`) as well
as a GPU-based delegated operator (`teams-do-with-gpu`):

```yaml
services:
  teams-do:
    extends:
      file: ../common-services.yaml

>       service: teams-do-common

Reviewer: This one will show up as "Builtin" in the UI

  teams-do-with-gpu:
    image: voxel51/fiftyone-teams-cv-full:v2.7.0
    deploy:
      replicas: ${FIFTYONE_DELEGATED_OPERATOR_WORKER_REPLICAS:-3}
    command: >

>       /bin/sh -c "fiftyone delegated launch -t remote"

Reviewer: ```suggestion
      /bin/sh -c "fiftyone delegated launch -t remote -n 'teams-do-with-gpu'"
```

Reviewer: if you don't add the -n I'm not sure how the app will deal with two different `Builtin` DOs...

Author: Ah, yes, thank you!

    environment:
      API_URL: ${API_URL}
      FIFTYONE_DATABASE_ADMIN: false
      FIFTYONE_DATABASE_NAME: ${FIFTYONE_DATABASE_NAME}
      FIFTYONE_DATABASE_URI: ${FIFTYONE_DATABASE_URI}
      FIFTYONE_ENCRYPTION_KEY: ${FIFTYONE_ENCRYPTION_KEY}
      FIFTYONE_INTERNAL_SERVICE: true

>       FIFTYONE_MEDIA_CACHE_SIZE_BYTES: -1

Author: I don't, this is just an example and follows what's in our base docker compose files. I think we should stay consistent, though.

>       FIFTYONE_MEDIA_CACHE_SIZE_BYTES: -1

Author: as a reminder, this doc isn't how to configure DO. This is just an example on how to use GPUs. It should be general and apply to plugins or any other workload

>       FIFTYONE_MEDIA_CACHE_SIZE_BYTES: -1

Reviewer 2: Yeah ok, I will follow up separately

<!-- markdownlint-disable no-inline-html line-length no-alt-text -->
<!-- markdownlint-disable-next-line first-line-heading -->
<div align="center">
<p align="center">

<img src="https://user-images.githubusercontent.com/25985824/106288517-2422e000-6216-11eb-871d-26ad2e7b1e59.png" height="55px"> &nbsp;
<img src="https://user-images.githubusercontent.com/25985824/106288518-24bb7680-6216-11eb-8f10-60052c519586.png" height="50px">

</p>
</div>
<!-- markdownlint-enable no-inline-html line-length no-alt-text -->

---

# Leveraging GPU Workloads

<!-- toc -->

- [Overview](#overview)
- [Utilizing GPUs For Delegated Operations](#utilizing-gpus-for-delegated-operations)
  - [Prerequisites](#prerequisites)
  - [Deploying GPU-enabled Delegated Operator Containers](#deploying-gpu-enabled-delegated-operator-containers)
  - [Validating GPU Access](#validating-gpu-access)

<!-- tocstop -->

## Overview

> In many machine learning applications, it is desirable to utilize available
> GPU hardware for intensive computations.

Reviewer 3: ```suggestion
Many machine learning applications utilize
GPU hardware for intensive computations.
```

> The FiftyOne Enterprise compose files allow users to schedule containers on
> GPU-enabled nodes using the docker compose
> [`deploy.resource.reservation.devices`][compose-deploy-resources]
> settings for individual services.

Reviewer 3: ```suggestion
The FiftyOne Enterprise docker compose files allow users to schedule containers on
GPU-enabled nodes using a service's
[`deploy.resource.reservation.devices`][compose-deploy-resources].
```

The below will show an example deploying a GPU-enabled container via docker
compose by following their
[documentation][compose-gpu-how-to]
in a FiftyOne Enterprise context.

## Utilizing GPUs For Delegated Operations

### Prerequisites

> This example assumes you have a docker compose node with GPU devices available.
> This example also assumes you have followed the
> [GPU prerequisites][compose-gpu-resources]
> outlined by docker.

Reviewer 3: ```suggestion
This example assumes you have a docker compose node with GPU devices available
and you have followed Docker's
[GPU prerequisites][compose-gpu-resources].
```

### Deploying GPU-enabled Delegated Operator Containers

> In this example, we will leverage GPUs for
> [delegated operators](./configuring-delegated-operators.md).

Reviewer 3: ```suggestion
We will configure the
[delegated operators](./configuring-delegated-operators.md)
with GPUs.
```

> Under `.services.teams-do` in your `compose.delegated-operators.yaml`,
> add the `deploy.resource.reservation.devices` configuration:

Reviewer 3: ```suggestion
In your `compose.delegated-operators.yaml`, under `.services.teams-do`, 
add the `deploy.resource.reservation.devices` configuration:
```

Starting with the file name helps the reader by supplying the context first, instead of in the middle.

> ```yaml
>   teams-do:

Reviewer 3: ```suggestion
```yaml
services:
  teams-do:
```

Since we are indenting, we mind as well include the top level key.

    extends:
      file: ../common-services.yaml
      service: teams-do-common
      deploy:
        resources:
          reservations:
            devices:
              - driver: nvidia
                count: 1
                capabilities: [gpu]
```

> Now redeploy your stack via `docker compose up -d` and wait for the

Reviewer 3: ```suggestion
Redeploy the stack via `docker compose up -d` and wait for the
```

`teams-do` containers to be deployed.

For advanced GPU configuration, including selecting specific GPU devices,
please refer to the

> [docker documentation][compose-gpu-how-to].

Reviewer 2: I'd propose adding a sentence like "To enable scheduling jobs on two distinct `teams-do` services, where one service is CPU-only and one is GPU-enabled, simply add another `teams-do` stanza. For instance, you might configure `teams-do-cpu` and `teams-do-gpu`.

Author: Multi-do isn't really a thing in compose. They would have to modify the compose files... that seems like something we wouldn't want to advertise. But for helm, we can definitely add that

Author: Eh, I guess we're doing it anyway...

### Validating GPU Access

> You can validate that the container can correctly access the GPU drivers using
> PyTorch's
> [cuda.is_available method][pytorch-cuda-is-available]
> from within the container.

Reviewer 3: ```suggestion
You may validate that the container can access the GPU drivers using
PyTorch's
[cuda.is_available method][pytorch-cuda-is-available]
by execing into the container and running
```

```shell
$ docker compose exec teams-do \
    python -c 'import torch; print(torch.cuda.is_available())'
True
```

> If `True` is printed, then you are ready to run computations on GPU hardware.

Reviewer 3: ```suggestion
If `True` is printed, then computations may run on GPU hardware.
```
